README
I started patching indent before 2.0 came out, sent my changes to the 2.0
maintainer who neither acknowledged nor incorporated the fixes.  Post-2.0
development of indent has not produced a version (through 2.2.7) which I find
useful since most of the differences I see between their respective outputs are
new bugs in the 2.x version.  And it still does not fix the bugs that I have
noticed myself in indent.  Perhaps the next maintainer of indent will know
something about regression testing -T.Dickey 2002/4/7

The original read-me follows:

-------------------------------------------------------------------------------

README for GNU `indent' 2.0, November, 1998

  GNU indent changes the appearance of a C or C++ program by inserting
  or deleting whitespace according to a plethora of options.  Some
  canned styles of formatting are supported as well.  GNU indent is a
  descendant of BSD indent.  This file is part of the GNU indent
  distribution.

  Read the file NEWS for more information, especially to see what's
  different from the last version, and what future versions may do.

  To install indent, please read the file INSTALL included with the
  distribution. If you are installing indent on VMS, see the file
  VMS-README.

  For license and copying information, see the file COPYING.

DOCUMENTATION

  For complete documentation on the Web, please visit:

     http://www.arceneaux.com/indent

  GNU uses a language called "texinfo" for its documentation source
  which produces both an on-line, interactive version and a hardcopy
  version for printing.  An HTML version is now produced as well.

  The on-line version, `indent.info', is installed along with the
  executable by typing "make install".  Refer to the Texinfo manual
  for more information on this system.

  The hardcopy version is produced from the source file `indent.texinfo'
  in as either a DVI or Postscript format file.  To produce the DVI
  version, type:

    make indent.dvi

  To produce the Postscript version, type:

    make indent.ps

  Note that you must have the TeX typographical system installed for
  this to work.

  The HTML version is included along with the distribution, and
  consists of the files:

     indent.html  indent_foot.html  indent_toc.html

  which must be installed in the same directory.  These files can also
  be accessed at the URL http://www.arceneaux.com/indent/manual.html

  To create the HTML version from the texinfo version, type:

     texi2html indent.texinfo

BUG REPORTS

  Please address bug-reports and suggestions or comments to:

    indent@arceneaux.com

  How to report bugs:

  Please begin your bug report with a concise and factual description
  of the behavior you consider to be aberrant.  If you have narrowed
  the problem down to the use of a particular option, please mention
  that.  In general, also be sure to mention the command line options
  used to invoke indent, as well as the contents of your .indent.pro
  if you are using one.

  Then include the *smallest possible* piece of code which generates
  the described symptom.  It is very helpful if you can manage to
  reduce the size of the sample code.

  Finally, if you have investigated the problem further, include
  debugger stack traces, modifications, or speculation you have.

SENDING CODE

  If you send a patch as part of your bug fix, or send me code for new
  features, please send them as diffs.  In particular, send me the
  result of "diff -c <old_file> <new_file>".

  Because of the regression testing I require of indent, it is
  extremely unlikely that your code will be installed into indent
  "as-is";  however it may still be helpful, and will likely have
  solved your particular problem.


NEWS
Release notes for GNU indent version 2.0, November 1998

* indent withstands the U. Wisconsin fuzz test. (See
  http://www.cs.wisc.edu/Dienst/UI/2.0/Describe/ncstrl.uwmadison/CS-TR-95-1268
  for details.)

* indent compiles with gcc -Wall and MSVC++ with no warnings.

* C++ code is handled to a great extent.  There is no special switch
  or option. However, there are still several cases which are not
  handled as well as they should be, and there are no options
  controlling C++ formatting specifically.

  There is only one case I'm aware of in which the output of indent
  breaks compilation;  this is nested templates such as:

    Aaa<int, Bbb<int> > anotherAB;

  are transformed into:

    Aaa < int, Bbb < int >>anotherAB;

  which is problematic for some compilers.  This is, in my opinion,
  the fault of the incredibly brain-dead design of C++, and it's not
  clear when or if indent will deal properly with this example.

* Wide strings and characters are now handled.

* The manual page, indent.1, has been removed from the distribution
  because it is not supported and is seriously out of date (it
  corresponds to indent 1.4).

* The "-bbb" option, which forces blanklines before boxed comments,
  has been re-instated.

* Trailing spaces are removed from output.

* Line-breaking behaviour has been improved.

* The troff option has been removed.

* You may now put either C ("/* ... */") or C++ ("// ...") comments in
  the indent profile.  Options within comments are ignored.

* The K&R style of leaving procedure names on the same line as their
  type is now better supported (the option "-psl").

* The "-l" option now works more consistently.  This may change some
  behaviour by breaking lines where they were not broken before.
  Specifying an argument of "0" will turn off this option.

* The unused options "-ps" and "-nps" were removed.

* When `indent' completes its work successfully, it exits with a value
  of 0.  Otherwise, one of the following values is returned:

  1 : This means `indent' was incorrectly invoked.
  2 : Errors occurred during formatting, but processing continued.
  3 : An unrecoverable error occurred during the processing of an input
      file.  If there are more input files, indent will proceed to
      process them, but will return the error value 3.
  4 : A serious internal error occurred.  All processing is terminated.
  5 : A system error during processing occurred.  In this case,
     `indent' terminates.

Release notes for GNU indent version 1.9.

* All reported bugs have been fixed.

* A new option "-lc" ("--comment-line-length") was added.

* Please read the section "BUG REPORTS" in the README file.

* Unless "interesting" bugs appear, this will be the last release of
  indent version 1.  The next release planned is version 2, which will
  (hopefully) integrate support for C++, and possibly Objective C.
  Additionally, a couple more OS's may be supported.

* A volunteer is now working on the `review' program, which examines C
  code and produces the options that GNU `indent' would use to produce
  code in that style.

Release Notes for GNU indent version 1.8.

* VMS and MS-DOS are now supported.  Thanks to MEHRDAD@glum.dev.cf.ac.uk
  and hnyman@lesti.hut.fi for their contributions.

  Please note that I have no way of testing this code, and with slight
  modifications, have installed it "as is".

  Also, note that these systems do not use `configure'.  See the file
  VMS-README for details of the VMS installation.  For the MS-DOS
  version, it is expected that the compiler will recognize the define
  __MSDOS__ as being set.

* C++ comments are now recognized.  More C++ support will be
  forthcoming, possibly along with support for Objective C.

* Formatting can be disabled for sections of code by placing that code
  between lines containing the comments /* *INDENT-OFF* */ and
  /* *INDENT-ON* */ (this also works in C++ comments).  See the info
  node "Disabling Formatting" for more info.

* The processing of comments has been changed.  See the info node on
  "Comments" for a description.

* The default (i.e., GNU) behaviour of struct handling has been
  slightly changed.

* All reported bugs have been fixed.


Release Notes for GNU indent version 1.7.

* All reported bugs have been fixed, except some of those dealing with
  comments.  I have completely rewritten the comment handling code in
  indent, but since this produces rather different comment formats
  (mostly, it makes several comment-related options actually work) I
  am waiting for the next release of indent to install this.

  To see what the differences in the comment handling will be, compare
  the file `comments.texinfo' to the section on comments in the info
  node for indent.

  It is possible that I have lost some bug reports, as some of my mail
  falls into a crack between two machines.  In the future, please be
  sure to use the address "bug-gnu-utils@prep.ai.mit.edu".

* All errors and warnings are output on the standard error stream,
  and conform to the GNU error message format.

* There is now more support for small (i.e., 16-bit) machines.

* There have been slight changes to the default behaviour, namely:

  - Column 1 comments are not formatted.

  - Preprocessor spaces are removed.

* When "-lps" is specified, whitespace preceding the '#' character is
  left untouched, as well as following whitespace.

* When the tabsize is specified as 1 or 0 (e.g., "-ts 0"), spaces are
  used instead of tabs.  There may be bugs in this behaviour, but they
  should vanish with the new comment code.


Release Notes for GNU indent version 1.6.

* All reported bugs have been fixed.

* The makefiles have been, hopefully, improved (see the file `README'
  for installation procedure).  Essentially, it is sufficient to type
  "make" to build indent.  `indent' is distributed with `configure'
  just as other GNU software.  Thus, in conformance with the GNU

gperf.c
/* *INDENT-OFF* */
/* ANSI-C code produced by gperf version 3.0.4 */
/* Command-line: gperf -LANSI-C -C -c -p -t -T -g -j1 -o -K rwd -N is_reserved indent.gperf  */
/* Computed positions: -k'1,3,5' */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
#error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gnu-gperf@gnu.org>."
#endif

#line 1 "indent.gperf"

/* Command-line: gperf -c -p -t -T -g -j1 -o -K rwd -N is_reserved indent.gperf  */

#define TOTAL_KEYWORDS 56
#define MIN_WORD_LENGTH 2
#define MAX_WORD_LENGTH 12
#define MIN_HASH_VALUE 2
#define MAX_HASH_VALUE 70
/* maximum key range = 69, duplicates = 0 */

#ifdef __GNUC__
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
hash (register const char *str, register unsigned int len)
{
  static const unsigned char asso_values[] =
    {
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71,  2, 71,  1, 14,  6,
       4, 10, 18,  0, 10,  0,  1,  3, 15, 33,
       6, 24, 19, 71,  7,  0,  0, 10, 47, 28,
      71, 71, 18, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71, 71, 71, 71, 71,
      71, 71, 71, 71, 71, 71
    };
  register unsigned hval =len;

  switch (hval)
    {
      default:
        hval += asso_values[(unsigned char)str[4]];
      /*FALLTHROUGH*/
      case 4:
      case 3:
        hval += asso_values[(unsigned char)str[2]];
      /*FALLTHROUGH*/
      case 2:
      case 1:
        hval += asso_values[(unsigned char)str[0]];
        break;
    }
  return hval;
}

#ifdef __GNUC__
__inline
#if defined __GNUC_STDC_INLINE__ || defined __GNUC_GNU_INLINE__
__attribute__ ((__gnu_inline__))
#endif
#endif
const struct templ *
is_reserved (register const char *str, register unsigned int len)
{
  static const struct templ wordlist[] =
    {
      {"", rw_none}, {"", rw_none},
#line 24 "indent.gperf"
      {"if", rw_sp_paren,},
#line 26 "indent.gperf"
      {"int", rw_decl,},
#line 23 "indent.gperf"
      {"goto", rw_break,},
#line 9 "indent.gperf"
      {"auto", rw_decl,},
#line 15 "indent.gperf"
      {"do", rw_sp_nparen,},
#line 34 "indent.gperf"
      {"static", rw_decl,},
#line 52 "indent.gperf"
      {"intptr_t", rw_decl,},
#line 54 "indent.gperf"
      {"intmax_t", rw_decl,},
#line 11 "indent.gperf"
      {"case", rw_case,},
#line 12 "indent.gperf"
      {"char", rw_decl,},
#line 36 "indent.gperf"
      {"switch", rw_switch,},
#line 47 "indent.gperf"
      {"sig_atomic_t", rw_decl,},
#line 17 "indent.gperf"
      {"else", rw_sp_nparen,},
#line 28 "indent.gperf"
      {"register", rw_decl,},
#line 32 "indent.gperf"
      {"signed", rw_decl,},
#line 13 "indent.gperf"
      {"const", rw_decl,},
#line 39 "indent.gperf"
      {"unsigned", rw_decl,},
#line 35 "indent.gperf"
      {"struct", rw_struct_like,},
#line 30 "indent.gperf"
      {"return", rw_return,},
#line 38 "indent.gperf"
      {"union", rw_struct_like,},
#line 29 "indent.gperf"
      {"restrict", rw_decl,},
#line 19 "indent.gperf"
      {"extern", rw_decl,},
#line 18 "indent.gperf"
      {"enum", rw_struct_like,},
#line 27 "indent.gperf"
      {"long", rw_decl,},
#line 49 "indent.gperf"
      {"size_t", rw_decl,},
#line 25 "indent.gperf"
      {"inline", rw_decl,},
#line 21 "indent.gperf"
      {"for", rw_sp_paren,},
#line 31 "indent.gperf"
      {"short", rw_decl,},
#line 37 "indent.gperf"
      {"typedef", rw_decl,},
#line 22 "indent.gperf"
      {"global", rw_decl,},
#line 10 "indent.gperf"
      {"break", rw_break,},
#line 48 "indent.gperf"
      {"ptrdiff", rw_decl,},
      {"", rw_none},
#line 16 "indent.gperf"
      {"double", rw_decl,},
#line 46 "indent.gperf"
      {"jmpbuf", rw_decl,},
#line 45 "indent.gperf"
      {"double_t", rw_decl,},
#line 60 "indent.gperf"
      {"wctrans_t", rw_decl,},
#line 14 "indent.gperf"
      {"default", rw_case,},
#line 57 "indent.gperf"
      {"clock_t", rw_decl,},
#line 58 "indent.gperf"
      {"time_t", rw_decl,},
#line 59 "indent.gperf"
      {"wint_t", rw_decl,},
#line 43 "indent.gperf"
      {"while", rw_sp_paren,},
#line 53 "indent.gperf"
      {"uintptr_t", rw_decl,},
#line 8 "indent.gperf"
      {"_Imaginary", rw_decl,},
#line 6 "indent.gperf"
      {"_Bool", rw_decl,},
#line 20 "indent.gperf"
      {"float", rw_decl,},
#line 33 "indent.gperf"
      {"sizeof", rw_sizeof,},
#line 44 "indent.gperf"
      {"float_t", rw_decl,},
#line 56 "indent.gperf"
      {"fpos_t", rw_decl,},
#line 41 "indent.gperf"
      {"void", rw_decl,},
#line 50 "indent.gperf"
      {"wchar_t", rw_decl,},
#line 7 "indent.gperf"
      {"_Complex", rw_decl,},
      {"", rw_none},
#line 61 "indent.gperf"
      {"wctype_t", rw_decl,},
#line 51 "indent.gperf"
      {"va_list", rw_decl,},
      {"", rw_none},
#line 55 "indent.gperf"
      {"uintmax_t", rw_decl,},
      {"", rw_none}, {"", rw_none},
#line 40 "indent.gperf"
      {"va_dcl", rw_decl,},
      {"", rw_none}, {"", rw_none}, {"", rw_none}, {"", rw_none}, {"", rw_none}, {"", rw_none}, {"", rw_none}, {"", rw_none},
#line 42 "indent.gperf"
      {"volatile", rw_decl,}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register int key = (int)hash (str, len);

      if (key <= MAX_HASH_VALUE && key >= 0)
        {
          register const char *s = wordlist[key].rwd;

          if (*str == *s && !strncmp (str + 1, s + 1, len - 1) && s[len] == '\0')
            return &wordlist[key];
        }
    }
  return 0;
}
/* *INDENT-ON* */
